File: Data-Structures-Algorithms-In-TS/AlgorithmsChapters/DP/GridUniqueWays.ts
function gridUniqueWays(m: number, n: number): number {
	const dp: number[][] = [];
	for (let i = 0; i < m; i++) {
		dp[i] = [];
		for (let j = 0; j < n; j++) {
			dp[i][j] = 0;
		}
	}

	dp[0][0] = 1;

	// Initialize first column.
	for (let i = 1; i < m; i++) {
		dp[i][0] = dp[i - 1][0];
	}

	// Initialize first row.
	for (let j = 1; j < n; j++) {
		dp[0][j] = dp[0][j - 1];
	}

	for (let i = 1; i < m; i++) {
		for (let j = 1; j < n; j++) {
			dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
		}
	}

	return dp[m - 1][n - 1];
}

function gridUnique3Ways(m: number, n: number): number {
	const dp: number[][] = [];
	for (let i = 0; i < m; i++) {
		dp[i] = [];
		for (let j = 0; j < n; j++) {
			dp[i][j] = 0;
		}
	}

	dp[0][0] = 1;

	// Initialize first column.
	for (let i = 1; i < m; i++) {
		dp[i][0] = dp[i - 1][0];
	}

	// Initialize first row.
	for (let j = 1; j < n; j++) {
		dp[0][j] = dp[0][j - 1];
	}

	for (let i = 1; i < m; i++) {
		for (let j = 1; j < n; j++) {
			dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j] + dp[i][j - 1];
		}
	}

	return dp[m - 1][n - 1];
}

console.log(gridUniqueWays(3, 3)); // 6
console.log(gridUnique3Ways(3, 3)); // 13

/*
 * 6
 * 13
 */

File: Data-Structures-and-Algorithms-using-Swift/AlgorithmsChapters/DP/GridUniqueWays.swift
import Foundation

func gridUniqueWays(_ m : Int, _ n : Int) -> Int {
	var dp : [[Int]] = Array(repeating: Array(repeating: 0, count: n), count: m)
	dp[0][0] = 1
	// Initialize first column.
	var i : Int = 1
	while (i < m) {
		dp[i][0] = dp[i - 1][0]
		i += 1
	}
	// Initialize first row.
	var j : Int = 1
	while (j < n) {
		dp[0][j] = dp[0][j - 1]
		j += 1
	}

	i = 1
	while (i < m) {
		j = 1
		while (j < n) {
			dp[i][j] = dp[i - 1][j] + dp[i][j - 1]
			j += 1
		}
		i += 1
	}
	return dp[m - 1][n - 1]
}

// Diagonal movement allowed.
func gridUnique3Ways(_ m : Int, _ n : Int) -> Int {
	var dp : [[Int]] = Array(repeating: Array(repeating: 0, count: n), count: m)
	dp[0][0] = 1
	
	// Initialize first column.
	var i : Int = 1
	while (i < m) {
		dp[i][0] = dp[i - 1][0]
		i += 1
	}

	// Initialize first row.
	var j : Int = 1
	while (j < n) {
		dp[0][j] = dp[0][j - 1]
		j += 1
	}

	i = 1
	while (i < m) {
		j = 1
		while (j < n) {
			dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j] + dp[i][j - 1]
			j += 1
		}
		i += 1
	}
	return dp[m - 1][n - 1]
}

// Testing code.
print(gridUniqueWays(3,3))
print(gridUnique3Ways(3,3))
/* 
6
13
 */
